icicle/backend/cpu/src/hash/cpu_poseidon.cpp

```diff
@@ -1,47 +1,295 @@
+      poseidon_constants[T].sparse_matrices =
+        poseidon_constants[T].pre_matrix + mds_matrix_len; // pre_matrix and mds_matrix have the same length.
+    }
+
+    s_cpu_backend_poseidon_constants_initialized = true;
     return eIcicleError::SUCCESS;
   }
 
-  REGISTER_POSEIDON_INIT_DEFAULT_CONSTANTS_BACKEND("CPU", cpu_poseidon_init_default_constants);
-
   template <typename S>
   class PoseidonBackendCPU : public HashBackend
   {
   public:
-    PoseidonBackendCPU(unsigned arity) : HashBackend("Poseidon-CPU", sizeof(S), arity * sizeof(S)) {}
+    PoseidonBackendCPU(unsigned t, const S* domain_tag)
+        : HashBackend("Poseidon-CPU", sizeof(S), sizeof(S) * (nullptr != domain_tag ? t - 1 : t)),
+          m_domain_tag(nullptr != domain_tag ? *domain_tag : S::zero()), m_use_domain_tag(nullptr != domain_tag), m_t(t)
+    {
+      init_default_constants();
+    }
 
+    // For merkle tree size should be equal to the arity of a single hasher multiplier by sizeof(S).
+    // For sponge function it could be any number.
     eIcicleError hash(const std::byte* input, uint64_t size, const HashConfig& config, std::byte* output) const override
     {
-      ICICLE_LOG_DEBUG << "Poseidon CPU hash() " << size << " bytes, for type " << demangle<S>()
-                       << ", batch=" << config.batch;
-      // TODO implement
+      unsigned int arity = m_use_domain_tag ? m_t - 1 : m_t;
+
+      // Currently sponge and padding functionalities are not supported.
+      if (size != arity * sizeof(S)) {
+        ICICLE_LOG_ERROR
+          << "Sponge function still isn't supported. The following should be true: (size == T) but it is not.\n";
+        return eIcicleError::INVALID_ARGUMENT;
+      }
+
+      // Call hash_single config.batch times.
+      for (int batch_hash_idx = 0; batch_hash_idx < config.batch; batch_hash_idx++) {
+        hash_single(input, output);
+        input += arity * sizeof(S);
+        output += sizeof(S);
+      }
+
+      return eIcicleError::SUCCESS;
+    }
+
+  private:
+    // This function performs a single hash according to parameters in the poseidon_constants[] struct.
+    eIcicleError hash_single(const std::byte* input, std::byte* output) const
+    {
+      const unsigned int T = m_t;
+
+      unsigned int alpha = poseidon_constants[T].alpha;
+      unsigned int nof_upper_full_rounds = poseidon_constants[T].nof_upper_full_rounds;
+      unsigned int nof_partial_rounds = poseidon_constants[T].nof_partial_rounds;
+      unsigned int nof_bottom_full_rounds = poseidon_constants[T].nof_bottom_full_rounds;
+      S* rounds_constants = poseidon_constants[T].rounds_constants;
+      S* mds_matrix = poseidon_constants[T].mds_matrix;
+      S* pre_matrix = poseidon_constants[T].pre_matrix;
+      S* sparse_matrices = poseidon_constants[T].sparse_matrices;
+      // Allocate temporary memory for intermediate calcs.
+      S* tmp_fields = new S[T];
```

Comment:
since T is small, we can maybe use stack memory instead (C array or preferably std::array). Should be faster compared to heap memory due to cache locality.
If you still prefer heap memory, note that modern C++ prefers smart pointers (to new/delete) so you don't have to release it explicitly, and also if an exception is thrown you don't leak.

Reply:
OK